clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1257,6 +1261,9 @@ public void flush() {
         this.sender.wakeup();
         try {
             this.accumulator.awaitFlushCompletion();
+            if (transactionManager != null) {
+                transactionManager.maybeClearLastError();
```

Comment:
We need to throw the error after clearing it.  This way we don't swallow the error that otherwise would've been thrown in commit.  This way:

- send (got error)
- commit

sequence fails if sends got any errors, and so does

- send (got error)
- flush
- commit

(with your change the latter sequence would just swallow the error, which makes it too easy to write buggy code for basic cases).

The advantage of new logic is that if someone wants to do advanced error handling and ignore some errors they'll be able to:

- send
- try { flush } catch { decide }
- commit

would work for advanced cases, without swallowing errors in basic cases.

Reply:
I thought the error is handed back via `Callback` and users are supposed to check for errors there? Why do we need to throw it?

Reply:
To make it harder to write buggy code by keeping semantics of similar constructs to be similar.  `send + flush + commit`  looks like just a more verbose form for `send + commit` , the latter works without handling errors in send (correct basic program with minimal effort), the former would swallow the error.